Olives: update to olives and emily 3.x, remove dependency on requirejs, Fixes #1029

podefr opened the pull request:
Hi,

This PR should fix #1029 (focus on load). The issue was fixed by updating olives and emily to their 3.0.x version.

The changes in this PR are:

1. Update to emily 3.0.3
2. Update to olives 3.0.5
3. Remove dependency on requirejs
4. Use node's require to replace requirejs
5. Use browserify to bundle the application
6. Update HTML to only load the bundle
7. Add a "build" and "watch" scripts to package.json to build the app
8. Use lowercase names for the UI components as they were never actually newed up.

This PR doesn't add the routing feature, this will come in a future PR as I'm working on it.

Thanks,
Olivier

arthurvr commented:
Thanks! I'll review this shortly.

arthurvr commented on examples/olives/package.json:
```diff
@@ -1,10 +1,13 @@
+  },
+  "scripts": {
+      "build": "browserify ./js/app.js -o olives-todo.js",
+      "watch": "watchify ./js/app.js -o olives-todo.js"
```
`package.json` files use 2 spaces indentation.

arthurvr commented on examples/olives/js/uis/list.js:
```diff
@@ -0,0 +1,77 @@
+        // Alive applies the plugins to the HTML view
+        list.alive(view);
+    };
+})();
```
Could you add a final newline?

arthurvr replied:
for the other files as well

arthurvr commented on examples/olives/js/lib/Tools.js:
```diff
@@ -1,20 +1,19 @@
+                this.classList.remove(className);
+            }
+        }
+    };
```
We use tabs indentation, could you fix that? Same applies to all other files except package.json

arthurvr replied:
Ping. It's still 4 spaces atm.

podefr replied:
ouch, never saw these two comments. I'm fixing this now.

arthurvr commented:
> This PR should fix #1029 (focus on load). The issue was fixed by updating olives and emily to their 3.0.x version.

Works for me in firefox, but not in Chrome or Safari. Could you double-check it?

arthurvr commented:
> This PR doesn't add the routing feature, this will come in a future PR as I'm working on it.

Thanks, awesome!

podefr commented:
Thanks for your feedback, I'll update this later on today and will push a new commit.

podefr commented:
I converted the spaces to tabs and added the newlines add the bottom of each JS file, although they don't seem to appear in github.
I double-checked in safari 8.0.4/chrome 43/firefox 31, and I do get the autofocus on the input field on load. Is anybody else seeing the same issue?

podefr commented:
oh and btw, let me know if you want me to squash the commits before merging

arthurvr commented:
LGTM at the moment. No idea what was going on with the autofocus as atm it works fine for me as well :/ All tests pass for olives except the routing. Awesome work @podefr! If nobody objects I'm gonna land this one.

arthurvr commented on examples/olives/package.json:
```diff
@@ -1,10 +1,13 @@
+    "olives": "^3.0.5",
+    "todomvc-app-css": "^1.0.1",
+    "todomvc-common": "^1.0.1"
+  },
```
Oh, one little thing, Browserify and watchify will need to be in the dev deps.

podefr replied:
good point, I'll fix it and push a new commit later on today

arthurvr replied:
@podefr Thanks

podefr commented:
I've added browserify and watchify to the package.json and squashed all the commits. Thanks for taking the time to review this @arthurvr!

arthurvr commented:
Great! LGTM.

samccone commented on examples/olives/js/lib/Tools.js:
```diff
@@ -1,4 +1,3 @@
-/*global define*/
 (function () {
```
the iffy wrapper seems odd now that we are using commonjs right?

sindresorhus replied:
:+1:

samccone commented:
I am not sure we want to delete

`examples/olives/node_modules/todomvc-app-css/index.css`

podefr commented:
@samccone, `index.css` hasn't been deleted, don't know why github is confused. It has been modified though, probably when updating the dependencies.

Here's what the file looks like as of this commit:

https://github.com/tastejs/todomvc/tree/92fdf4fbd67132df26ee468d46304bcbe9cbad88/examples/olives/node_modules/todomvc-app-css

podefr commented:
@arthurvr @samccone I believe I've answered all your comments, let me know if there's anything else I can do so you can close this, thanks.

arthurvr commented on examples/olives/js/lib/Tools.js:
```diff
@@ -1,20 +1,17 @@
+        } else {
+            this.classList.remove(className);
+        }
+    }
```
For this file same thing as https://github.com/tastejs/todomvc/pull/1242#discussion_r27489362

arthurvr replied:
And it looks like all the other files as well. package.json is the only one that shouldn't use tabs.

arthurvr commented on examples/olives/js/lib/Tools.js:
```diff
@@ -1,20 +1,17 @@
+ */
+module.exports = {
+    // className is set to the 'this' dom node according to the value's truthiness
+    'toggleClass': function (value, className) {
```
Unrelated to this PR but perfect time to change it, can you remove the quotes?

podefr commented:
@arthurvr done!

samccone commented:
hi @podefr looks like you have a failing spec

```
 TodoMVC - olives
    When page is initially opened
      ✓ should focus on the todo input field
    No Todos
      ✓ should hide #main and #footer (74ms)
    New Todo
      ✓ should allow me to add todo items (381ms)
      ✓ should clear text input field when an item is added (175ms)
      ✓ should append new items to the bottom of the list (468ms)
      ✓ should trim text input (213ms)
      ✓ should show #main and #footer when items added (221ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (494ms)
      ✓ should allow me to clear the completion state of all items (556ms)
      ✓ complete all checkbox should update state when items are completed / cleared (585ms)
    Item
      ✓ should allow me to mark items as complete (480ms)
      ✓ should allow me to un-mark items as complete (441ms)
      ✓ should allow me to edit an item (771ms)
      ✓ should show the remove button on hover
    Editing
      ✓ should hide other controls when editing (443ms)
      ✓ should save edits on enter (666ms)
      ✓ should save edits on blur (847ms)
      ✓ should trim entered text (818ms)
      ✓ should remove the item if an empty text string was entered (634ms)
      ✓ should cancel edits on escape (735ms)
      1) "after each" hook

  20 passing (45s)
  1 failing

  1) TodoMVC - olives "after each" hook:
     Uncaught AssertionError: 3 items expected in the todo list, 2 items observed
      at /Users/sam/Desktop/repos/todomvc/browser-tests/testOperations.js:126:12
      at _fulfilled (/Users/sam/Desktop/repos/todomvc/browser-tests/node_modules/q/q.js:794:54)
      at self.promiseDispatch.done (/Users/sam/Desktop/repos/todomvc/browser-tests/node_modules/q/q.js:823:30)
      at Promise.promise.promiseDispatch (/Users/sam/Desktop/repos/todomvc/browser-tests/node_modules/q/q.js:756:13)
      at /Users/sam/Desktop/repos/todomvc/browser-tests/node_modules/q/q.js:564:44
      at flush (/Users/sam/Desktop/repos/todomvc/browser-tests/node_modules/q/q.js:110:17)
      at process._tickCallback (node.js:419:13)
```

podefr commented:
Hey, it seems that this isn't new right? It doesn't appeared to be caused by me fixing #1029, so can we close this PR, create a new defect, and I'll push a new code review?

thanks!

podefr commented:
@samccone I pulled down the latest master and ran:

```bash
npm run test -- --framework=olives
```

but I'm not seeing any failing test apart from routing. Can you double-check?

```
   TodoMVC - olives
    When page is initially opened
      ✓ should focus on the todo input field
    No Todos
      ✓ should hide #main and #footer (79ms)
    New Todo
      ✓ should allow me to add todo items (313ms)
      ✓ should clear text input field when an item is added (145ms)
      ✓ should append new items to the bottom of the list (336ms)
      ✓ should trim text input (149ms)
      ✓ should show #main and #footer when items added (185ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (366ms)
      ✓ should allow me to clear the completion state of all items (397ms)
   TodoMVC - olives
    When page is initially opened
      ✓ should focus on the todo input field
    No Todos
      ✓ should hide #main and #footer (73ms)
    New Todo
      ✓ should allow me to add todo items (303ms)
      ✓ should clear text input field when an item is added (143ms)
      ✓ should append new items to the bottom of the list (347ms)
      ✓ should trim text input (154ms)
      ✓ should show #main and #footer when items added (185ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (369ms)
      ✓ should allow me to clear the completion state of all items (381ms)
      ✓ complete all checkbox should update state when items are completed / cleared (424ms)
    Item
      ✓ should allow me to mark items as complete (375ms)
      ✓ should allow me to un-mark items as complete (326ms)
      ✓ should allow me to edit an item (524ms)
      ✓ should show the remove button on hover
    Editing
      ✓ should hide other controls when editing (318ms)
      ✓ should save edits on enter (521ms)
      ✓ should save edits on blur (612ms)
      ✓ should trim entered text (533ms)
      ✓ should remove the item if an empty text string was entered (471ms)
      ✓ should cancel edits on escape (449ms)
    Counter
      ✓ should display the current number of todo items (232ms)
    Clear completed button
      ✓ should display the number of completed items (375ms)
      ✓ should remove completed items when clicked (403ms)
      ✓ should be hidden when there are no items that are completed (369ms)
    Routing
      1) should allow me to display active items
      2) should respect the back button
      3) should allow me to display completed items
      4) should allow me to display all items
      5) should highlight the currently applied filter


  24 passing (44s)
  5 failing

```

podefr commented:
@samccone @arthurvr seems like the tests are passing for me (expect for routing), can you double check and maybe close this PR that's been open for 20 days? I'll add routing as part of another PR.

Thanks,
Olivier

arthurvr commented:
@podefr Indentation is still wrong. Otherwise I'm fine with landing this as olives doesn't pass the specs in master too.

podefr commented:
thanks @arthurvr, fixing the indent now. I had never saw the comment above about using tabs instead of spaces :/

podefr commented:
@arthurvr done

arthurvr commented:
@samccone I can actually confirm the failing test, but the spec which is failing is due to https://github.com/tastejs/todomvc/issues/789, which is another issue and thus not topic for this PR.

arthurvr commented on examples/olives/package.json:
```diff
@@ -1,10 +1,17 @@
+  "scripts": {
+    "build": "browserify ./js/app.js -o olives-todo.js",
+    "watch": "watchify ./js/app.js -o olives-todo.js"
+  },
```
Mind adding a note about these scripts in the `examples/olives/readme.md`? Thanks!

podefr replied:
will do

podefr replied:
done

arthurvr commented:
one last comment, for the rest this looks good.

arthurvr commented:
Oh, and, thanks for your hard work!

samccone commented:
:clap:

podefr commented:
@arthurvr @samccone I've updated the README with the notes about the two scripts. When this is merged in, I'll add the router

arthurvr commented on examples/olives/readme.md:
````diff
@@ -16,3 +16,25 @@ Here are some links you may find helpful:
+npm install
+```
+
+
````
remove these lines :)

arthurvr commented:
Instead of nitpicking more I'll land this this and do some tweaks myself! Once again, awesome work!

arthurvr commented:
> When this is merged in, I'll add the router

That would be super awesome!
